src/charm.py

```diff
@@ -140,349 +154,143 @@ def __init__(self, *args):
-        """Define and start a workload using the Pebble API."""
-        self.unit.open_port(protocol="tcp", port=ADMIN_UI_PORT)
-
+        self._workload_service.open_port()
         self._handle_status_update_config(event)
 
-        if not self._container.can_connect():
-            event.defer()
-            logger.info("Cannot connect to admin-ui container. Deferring the event.")
-            self.unit.status = WaitingStatus("Waiting to connect to admin-ui container")
-            return
-
-        self._set_version()
+        version = self._workload_service.version
+        self._workload_service.version = version
 
     def _on_config_changed(self, event: ConfigChangedEvent) -> None:
-        """Handle changed configuration."""
         self._handle_status_update_config(event)
 
     def _on_peer_relation_changed(self, event: RelationChangedEvent) -> None:
         self._handle_status_update_config(event)
 
-    def _on_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
-        if self.unit.is_leader():
-            logger.info("This app's public ingress URL: %s", event.url)
-
-    def _on_ingress_revoked(self, event: IngressPerAppRevokedEvent) -> None:
-        if self.unit.is_leader():
-            logger.info("This app no longer has ingress")
-
+    @wait_when(
+        container_not_connected,
+        integration_not_exists(PEER_INTEGRATION_NAME),
+    )
     def _on_openfga_store_created(self, event: OpenFGAStoreCreateEvent) -> None:
-        """Handle openfga store created event."""
-        if not self._container.can_connect():
-            event.defer()
-            logger.info("Cannot connect to admin-ui container. Deferring the event.")
-            self.unit.status = WaitingStatus("Waiting to connect to admin-ui container")
-            return
-
-        if not self._peers:
-            self.unit.status = WaitingStatus("Waiting for peer relation")
+        if not self.openfga_integration.is_store_ready():
             event.defer()
             return
 
-        openfga_info = self._get_openfga_store_info()
-        if not openfga_info:
-            logger.debug("No openfga store info found, deferring the event")
-            event.defer()
-            return
+        if self.unit.is_leader():
+            openfga_model_id = self._workload_service.create_openfga_model(
+                self.openfga_integration.openfga_integration_data
+            )
+            self.peer_data[self._workload_service.version] = {"openfga_model_id": openfga_model_id}
 
-        self._create_openfga_model(openfga_info)
         self._handle_status_update_config(event)
 
     def _on_openfga_store_removed(self, event: OpenFGAStoreRemovedEvent) -> None:
-        """Handle openfga store removed event."""
-        logger.info("OpenFGA store was removed")
-        if self.unit.is_leader():
-            self._pop_peer_data(key=self._get_version())
-
+        # TODO: need to check if leader unit is a must
```

Comment:
comment: I'd say yes, but I'll let the rest of the team speak on this

Reply:
Yeah, this is a must. Will put on the leader unit check later.